shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/TemporaryTerminalFilter.kt

```diff
@@ -0,0 +1,136 @@
+import com.mbta.tid.mbta_app.model.response.PredictionsStreamDataResponse
+import com.mbta.tid.mbta_app.model.response.ScheduleResponse
+
+/**
+ * When part of a subway line is closed, should the headsign be the temporary terminal or the
+ * permanent terminal? There is no obviously correct answer - temporary terminals directly
+ * communicate the boundaries of a disruption, but permanent terminals match in-station wayfinding
+ * and are more clearly aligned with directions for riders unfamiliar with the system.
+ *
+ * The data we receive is inconsistent: depending on numerous factors upstream, schedules may or may
+ * not contain temporary terminals, and if they do, predictions may or may not match them. In the
+ * medium term, we expect to predictably receive permanent terminals as headsigns, but in the short
+ * term, we need some way to avoid confusion in the presence of inconsistent data.
+ *
+ * Rewriting predictions to have different headsigns and route patterns was a mess when we tried it
+ * previously, so we always keep predictions as-is, whether they show permanent terminals or
+ * temporary, and if the schedules disagree, we erase the scheduled headsign. Since we don't show
+ * scheduled times on subway, the hidden schedule headsign would only ever be shown with
+ * "Predictions unavailable", so no actual data will be lost.
+ */
+class TemporaryTerminalFilter(
+    val nearbyStaticData: NearbyStaticData,
+    val predictions: PredictionsStreamDataResponse,
+    val globalData: GlobalResponse,
+    val alerts: List<Alert>,
+    val schedules: ScheduleResponse
+) {
+    private val schedulesByRoute = schedules.schedules.groupBy { it.routeId }
+    // predictions that reflect auto-cancelled schedules will throw everything off, so filter out
+    // cancellations
+    private val predictionsByRoute =
+        predictions.predictions.values
+            .filterNot { it.scheduleRelationship == Prediction.ScheduleRelationship.Cancelled }
+            .groupBy { it.routeId }
+    private val routePatternsByRoute = globalData.routePatterns.values.groupBy { it.routeId }
+
+    fun Schedule.routePatternId() = schedules.trips[tripId]?.routePatternId
+
+    fun Prediction.routePatternId() = predictions.trips[tripId]?.routePatternId
+
+    /**
+     * Rewriting realtime data for temporary terminals causes what the app displays to diverge from
+     * what the API says is real, and so must be done with caution. Specifically, we only want to
+     * rewrite
+     * 1. subway routes
+     * 2. with alerts somewhere on the line
+     * 3. where the schedule has non-typical patterns and is missing a typical pattern (requiring
+     *    only non-typical patterns misses single-branch RL disruptions, but allowing just any
+     *    non-typical pattern hits GL early-morning cross-branch trips)
+     * 4. but the predictions all have typical patterns instead of non-typical patterns (meaning
+     *    there are no predictions that will be lost)
```

Comment:
I was thinking that we still need 3 to avoid hitting the Green Line early-morning cross-branch trips as mentioned in the comment, because even if there is an alert elsewhere on the line it would be incorrect to hide those headsigns before the predictions load in, but if the alert is scheduled then there will be a non-typical pattern replacing a typical pattern to reflect that disruption, so 3 only makes a difference for unscheduled disruptions, and we could remove that check now in the name of simplicity if we decided it was worth doing. I don't think we need 4, though.

Reply:
>  it would be incorrect to hide those headsigns before the predictions load in
Now that we have shimmer loading, is that still relevant?

If it is easy to get rid of 3 and/or for I'd be in favor of it for simplicity should this code stick around longer than we'd like, but it is planned to be short-lived, so I'm OK with is staying.